src/algebra/commute.lean

```diff
@@ -0,0 +1,230 @@
+/-
+Copyright (c) 2019 Neil Strickland. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Neil Strickland
+
+Commuting pairs of elements in not-necessarily-commutative monoids,
+groups and (semi) rings.  Centralizers as submonoids, subgroups or
+subrings.
+-/
+
+import algebra.group_power
+  group_theory.submonoid group_theory.subgroup
+  ring_theory.subring
+
+section monoid
+
+variables {M : Type*} [monoid M]
+
+@[reducible] def commute (a b : M) : Prop := a * b = b * a
+
+theorem commute.eq {a b : M} (h : commute a b) : a * b = b * a := h
+
+theorem commute.refl (a : M) : commute a a := @rfl M (a * a)
+
+theorem commute.symm {a b : M} : commute a b → commute b a :=
+λ h, h.symm
+
+theorem commute.one (a : M) : commute a 1 :=
+by { dsimp [commute], rw [one_mul, mul_one] }
+
+theorem commute.one_left (a : M) : commute 1 a := (commute.one a).symm
+
+theorem commute.mul {a b c : M} :
+ commute a b → commute a c → commute a (b * c) :=
+λ hab hac,
+  by { dsimp [commute] at *,
+       rw [mul_assoc, ← hac, ← mul_assoc b, ← hab, mul_assoc] }
+
+theorem commute.mul_left {a b c : M} :
+ commute a c → commute b c → commute (a * b) c :=
+λ hac hbc, (hac.symm.mul hbc.symm).symm
+
+theorem commute.pow {a b : M} (hab : commute a b) : ∀ (n : ℕ), commute a (b ^ n)
+| 0 := by { rw [pow_zero], exact commute.one a }
+| (n + 1) := by { rw [pow_succ], exact hab.mul (commute.pow n) }
+
+theorem commute.pow_left {a b : M} (hab : commute a b) (n : ℕ) : commute (a ^ n) b :=
+(hab.symm.pow n).symm
+
+theorem commute.pow_pow {a b : M} (hab : commute a b) (n m : ℕ) :
+ commute (a ^ n) (b ^ m) :=
+commute.pow (commute.pow_left hab n) m
+
+theorem commute.self_pow (a : M) (n : ℕ) : commute a (a ^ n) :=
+(commute.refl a).pow n
+
+theorem commute.pow_self (a : M) (n : ℕ) : commute (a ^ n) a :=
+(commute.refl a).pow_left n
+
+theorem commute.pow_pow_self (a : M) (n m : ℕ) : commute (a ^ n) (a ^ m) :=
+(commute.refl a).pow_pow n m
+
+def centralizer (a : M) : set M := λ x, commute a x
+
+theorem mem_centralizer (a : M) {x : M} : x ∈ centralizer a ↔ commute a x :=
+by { dsimp [centralizer], refl }
+
+instance centralizer_submonoid (a : M) : is_submonoid (centralizer a) :=
```

Comment:
I would rather call this `centralizer.is_submonoid`. That is the name that Lean would give this if you put it in the `centralizer` namespace, and didn't specify the name. If/when at some point we teach type class inference to guess the correct name during tc search, we will be glad if we follow the auto-naming convention as much as possible.

Reply:
Done.